code/modules/vote/vote_datum.dm

```diff
@@ -0,0 +1,192 @@
+#define VOTE_RESULT_TYPE_MAJORITY "Majority"
+
+/datum/vote
+	/// Person who started the vote
+	var/initiator = "the server"
+	/// world.time the vote started at
+	var/started_time = null
+	/// The question being asked
+	var/question = null
+	/// Vote type text, for showing in UIs and stuff
+	var/vote_type_text = "unset"
+	/// Do we want to show the vote counts as it goes
+	var/show_counts = FALSE
+	/// Vote result type. This determines how a winner is picked
+	var/vote_result_type = VOTE_RESULT_TYPE_MAJORITY
+	/// Was this vote custom started?
+	var/is_custom = FALSE
+	/// Choices available in the vote
+	var/list/choices = list()
+	// Assoc list of [ckeys => choice] who have voted. We dont want to hold client refs.
+	var/list/voted = list()
+
+
+/datum/vote/New(_initiator, _question, list/_choices, _is_custom = FALSE)
+	if(SSvote.active_vote)
+		CRASH("Attempted to start another vote with one already in progress!")
+
+	if(_initiator)
+		initiator = _initiator
+	if(_question)
+		question = _question
+	if(_choices)
+		choices = _choices
+
+	is_custom = _is_custom
+
+	// If we have no choices, dynamically generate them
+	if(!length(choices))
+		generate_choices()
+
+	var/text = "[capitalize(vote_type_text)] vote started by [initiator]."
+	if(is_custom)
+		vote_type_text = "custom"
+		text += "\n[question]"
+		if(usr)
+			log_admin("[capitalize(vote_type_text)] ([question]) vote started by [key_name(usr)].")
+
+	else if(usr)
+		log_admin("[capitalize(vote_type_text)] vote started by [key_name(usr)].")
+
+	log_vote(text)
+	started_time = world.time
+	SSvote.active_vote = src
+	announce(text)
+
+/datum/vote/proc/remaining()
+	return max(((started_time + GLOB.configuration.vote.vote_time) - world.time), 0)
+
+
+// Returns the result
+/datum/vote/proc/calculate_result()
+	switch(vote_result_type)
+		if(VOTE_RESULT_TYPE_MAJORITY)
+			var/list/results = list()
+
+			// Count up all votes
+			for(var/ck in voted)
+				if(voted[ck] in results)
+					results[voted[ck]]++
+				else
+					results[voted[ck]] = 1
+
+			// Get the biggest vote count, since we can also use this to pick tiebreaks
+			var/maxvotes = 0
+			for(var/res in results)
+				if(results[res] > maxvotes)
+					maxvotes = results[res]
+
+			var/list/winning_options = list()
+
+			for(var/res in results)
+				if(results[res] == maxvotes)
+					winning_options |= res
+
+			if(!length(winning_options))
```

Comment:
this is a bit of a roundabout way to check the absence of votes, which relies on *not* doing an otherwise reasonable `for(var/choice in choices) results[choice] = 0` to pre-populate the counting boxes.

I think changing this to `if(!length(voted))` and moving it to the start of the block would be a more natural way to check for this.